device/core/src/internal_client.ts

```diff
@@ -642,6 +643,105 @@ export class InternalClient extends EventEmitter {
+     *                      IOTEDGE_WORKLOADURI        - URI for iotedged's workload API
+     *                      IOTEDGE_DEVICEID           - Device identifier
+     *                      IOTEDGE_MODULEID           - Module identifier
+     *                      IOTEDGE_MODULEGENERATIONID - Module generation identifier
+     *                      IOTEDGE_IOTHUBHOSTNAME     - IoT Hub host name
+     *                      IOTEDGE_AUTHSCHEME         - Authentication scheme to use;
+     *                                                   must be "SasToken"
+     *
+     * @param {Function}  transportCtor     A transport constructor.
+     * @param {Function}  clientCtor        The client object constructor.
+     */
+  static fromEnvironment(transportCtor: any, clientCtor: any): any {
+    // Codes_SRS_NODE_INTERNAL_CLIENT_13_026: [ The fromEnvironment method shall throw a ReferenceError if the transportCtor argument is falsy. ]
+    if (!transportCtor) {
+      throw new ReferenceError('transportCtor cannot be \'' + transportCtor + '\'');
+    }
+
+    // Codes_SRS_NODE_INTERNAL_CLIENT_13_027: [ The fromEnvironment method shall throw a ReferenceError if the clientCtor argument is falsy. ]
+    if (!clientCtor) {
+      throw new ReferenceError('clientCtor cannot be \'' + clientCtor + '\'');
+    }
+
+    // Codes_SRS_NODE_INTERNAL_CLIENT_13_028: [ The fromEnvironment method shall delegate to InternalClient.fromConnectionString if an environment variable called EdgeHubConnectionString or IotHubConnectionString exists. ]
+
+    // if the environment has a value for EdgeHubConnectionString then we use that
+    const connectionString = process.env.EdgeHubConnectionString || process.env.IotHubConnectionString;
+    if (connectionString) {
+      return InternalClient.fromConnectionString(
+        connectionString,
+        transportCtor,
+        clientCtor
+      );
+    }
+
+    // make sure all the environment variables we need have been provided
+    InternalClient.validateEnvironment();
+
+    const authConfig = {
+      workloadUri: process.env.IOTEDGE_WORKLOADURI,
+      deviceId: process.env.IOTEDGE_DEVICEID,
+      moduleId: process.env.IOTEDGE_MODULEID,
+      iothubHostName: process.env.IOTEDGE_IOTHUBHOSTNAME,
+      authScheme: process.env.IOTEDGE_AUTHSCHEME,
+      gatewayHostName: process.env.IOTEDGE_GATEWAYHOSTNAME,
+      generationId: process.env.IOTEDGE_MODULEGENERATIONID
+    };
+
+    // Codes_SRS_NODE_INTERNAL_CLIENT_13_032: [ The fromEnvironment method shall create a new IotEdgeAuthenticationProvider object and pass this to the transport constructor. ]
+    const authenticationProvider = new IotEdgeAuthenticationProvider(authConfig);
+
+    // Codes_SRS_NODE_INTERNAL_CLIENT_13_031: [ The fromEnvironment method shall return a new instance of the Client object. ]
+    return new clientCtor(
+      new transportCtor(authenticationProvider),
+      null,
+      new BlobUploadClient(authenticationProvider)
+    );
+  }
 }
```

Comment:
I think that all of this belongs only in the module_client file, and that the ModuleClient class could call only internalClient.fromConnectionString or internalClient.fromAuthenticationProvider?

Reply:
Sure. Will move.